### ISSUE TITLE
Why does Lst<T> not override ToString?

### ISSUE THREAD

**Issue by gonzaw** [louthy/language-ext#593]
Hi. This is more of a question about the API.

Why doesn't Lst<T> override ToString? For example, I'd like to be able to use Lst<T> in records and be able to get a human-readable view of all its contents when I call "ToString" on the record, including the contents of the list.

Thanks!

**Comment by louthy** [comment]
@gonzaw I have thought about it, it's just when the list gets large then it can freak out in the debugger.  I realise that's not a great way to make a decision about functionality, but it's relatively easy to do:
```c#
   var str = String.Join(", ", list);
```
So it's not such a huge problem.

**Comment by gonzaw** [comment]
> @gonzaw I have thought about it, it's just when the list gets large then it can freak out in the debugger. I realise that's not a great way to make a decision about functionality, but it's relatively easy to do:
> 
> ```cs
>    var str = String.Join(", ", list);
> ```
> 
> So it's not such a huge problem.

What if I have a datatype like this?

    public class MyDataType : Record<MyDataType>
    {
        public readonly int Value1;
        public readonly string Value2;
        public readonly Lst<int> Value3;

        public MyDataType(int value1, string value2, Lst<int> value3)
        { 
            this.Value1 = value1;
            this.Value2 = value2;
            this.Value3 = value3;
        }
    }

    MyDataType data = ... // Get the value from somewhere else external
    var str = data.ToString();

How can I make it so `str` shows "MyDataType(10, Text, [1, 2, 3])" for example?

**Comment by bender2k14** [comment]
Try `RecordType<MyDataType>.ToString(data)`

**Comment by louthy** [comment]
@bender2k14 That will use the `Lst` implementation of `ToString()` which does nothing.  

Personally I'm not against the idea of implementing `ToString()` for any of the collection types, it's the right thing to do, but it does open some problems for large lists.  I think often the reason for using `ToString()` is either:

1. Debugging
2. Serialisation

For 1 it's problematic if you have 10,000,000 items; for 2 you should probably be doing it yourself anyway.

**Comment by gonzaw** [comment]
Can't the implementation of ToString use StringBuilder and have a configuration that states the maximum amount of bytes/characters the resulting string can get? If when constructing the string representation of the list that limit is breached, you can just truncate it "[1, 2, 3, 99,999,998, 9,999,999, ... ]".

Would that work? There can also be a different ToString method exposed in the API that prints the whole string (if someone does want to print EVERYTHING, even if it surpasses the limit). E.g "Lst.ToFullString()"

The main issue is the broken contract and inconsistency IMO. If you do choose to use ToString, it's completely broken because it does not do what you want it to do with lists/collections. So ... why does ToString exist at all then? Not only that, but there is no canonical `IShowable` alternative (like in Haskell) for quick printing of values (for debugging, for logging, for REPLing/testing, and more use cases).

In this specific case of record, maybe even have a `RecordType<MyDataType>.ToFullString(data)` that does this (while leaving `Object.ToString()` alone). This would work for my specific use case (I'm using a logging framework that at some point calls `Object.ToString` on a dictionary of custom values that you can pass it, and it logs those values. Using JSON.NET or serialization is overkill, it forces me to add dependencies to the serialization library in every single method in my codebase, it´s not really an option)

**Comment by StefanBertels** [comment]
My 2 pence: 

1 .Implementing `ToString()` in some way will raise the problem that someone might use this in code other than log/debug outputs. So thinking twice is a good idea here. My personal opinion is to avoid using `ToString()` in code other than log or debug output.

2. I often have generic code and use `ToString()` for logging. The result for e.g. `HashMap` is quite useless. Would be nice all LanguageExt types would give output useful in debugging. It already works for simple types like `Option` in a good way and I don't expect this output to be stable (between releases). I think types containing mass data (Lst, Map, Arr, ...) should just print something human readable without performance issues (evaluate fast) and without size problems (just show the count and the first 3 and the last 1 or 3 items, maybe even dynamic). 

3. Question: Should the type information always be dropped?

4. One could implement `ToString()` with `Formattable` so a user could select one of multiple formats. I experimented with this for `Option` (because I wanted something like the output of Nullables). Might be very useful in string interpolation. But of course supporting multiple formats might open the box of pandora... Nevertheless one could select a simple scheme like: default output is like the output of `Option.ToString()` (but shortened for e.g. lists) and there could be alternatives in form of a format for "output plain value" (like `Option.ToNullable()`) and/or for  "output type information" (like current HashMap.ToString) and/or "output value with type information" (default plus type information). Anyway I would think that any option should just be used for log/debugging (which is a contra argument for "output plain value").

**Comment by StefanBertels** [comment]
Another note regarding data types with lazy evalution: Should `ToString()` evaluate? 

I think there cannot be no (complete) evaluation of `Seq` because it might enumerate an endless `IEnumerable`... Perhaps there should be no evaluation of any lazy operation. If we output content (like first three items or similar) in case of already evaluated lazy things then this means output `ToString()` will vary dependent on inner state. I personally think this is not a bad thing (useful for debugging, especially).

**Comment by mnn** [comment]
It's pretty bad for testing with NUnit:

```csharp
      Assert.AreEqual(Some(List(1)), Solution02.ParseInput("1\n"));
```

```csharp
Message: 
  Expected and actual are both <LanguageExt.Option`1[LanguageExt.Lst`1[System.Int32]]>
```
The message is useless.

If `toString` would be implemented, it might save a bit of time with debugging, because it would be obvious at glance what's wrong:

```csharp
Message: 
  Expected: Some(List(1))
  But was:  Some(List(2))
```

**Comment by StefanBertels** [comment]
Yes, would improve test output. But be aware that you probably* won't see a difference if just element 5000 in a 10000 element list differs, neither.

*(toString hopefully won't display whole list)

If you want to make sure to get the difference in test output you can use something like "a without b must be empty" and "b without a must be empty".

**Comment by mnn** [comment]
> *(toString hopefully won't display whole list)

Curious, why output of `toString` *should be* truncated? Shouldn't this responsibility be on user/tool, rather than general `toString` method?

For example Haskell nor Scala does this (Scala, or rather sbt, does truncate output in repl, but `toString` returns every element).

**Comment by mnn** [comment]
> Perhaps there should be no evaluation of any lazy operation. If we output content (like first three items or similar) in case of already evaluated lazy things then this means output ToString() will vary dependent on inner state. I personally think this is not a bad thing (useful for debugging, especially).

Tested this on Scala's stream and it is exactly as what you propose:

```scala
scala> val x = List(1, 2, 3).toStream
x: scala.collection.immutable.Stream[Int] = Stream(1, ?)

scala> x.toString
res9: String = Stream(1, ?)

scala> x(1)
res10: Int = 2

scala> x.toString
res11: String = Stream(1, 2, ?)
```

Haskell on the other hand doesn't mind printing "infinite" list. I think Scala is more similar to this library (C# being OOPL and so on).

**Comment by louthy** [comment]
Released in [`3.3.49`](https://github.com/louthy/language-ext/releases/tag/v3.3.49)
